### ISSUE TITLE
Passing containers by reference broken by #1455

### ISSUE THREAD

**Issue by clauniel** [pybind/pybind11#1465]
Normally I have been able to pass containers to the python interpreter in embedded interpreter mode and modify the objects inside this vector from python. PR #1455 breaks this.

Example code:

CMakeLists.txt
  
```cmake
cmake_minimum_required(VERSION 3.10)
project(PyBindRegression)
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fvisibility=hidden")

find_package(pybind11 REQUIRED)

set(TEST_SRC_FILES main.cpp)

add_executable(example ${TEST_SRC_FILES})
target_link_libraries(example pybind11::embed)

```

main.cpp
```C++
#include <pybind11/embed.h>
#include <pybind11/stl_bind.h>
#include <vector>
#include <iostream>

class A {
public:
    void setData(int data) { this->data = data; }
    int getData() const { return data; }
private:
    int data = 0;
};

class B {
public:
    B() : AVec(std::vector<A>(10)) {}
    std::vector<A>& getAVec() { return AVec; }
private:
    std::vector<A> AVec;
};

PYBIND11_MAKE_OPAQUE(std::vector<A>)

PYBIND11_EMBEDDED_MODULE(mod, m) {
    pybind11::class_<A>(m, "A")
	.def("get_data", &A::getData)
	.def("set_data", &A::setData);

    pybind11::bind_vector<std::vector<A>>(m, "AVector");

    pybind11::class_<B, std::shared_ptr<B>>(m, "B")
	.def("get_A_vec", &B::getAVec, pybind11::return_value_policy::reference);
}

int main() {
    pybind11::scoped_interpreter interp{};

    auto myB = std::make_shared<B>();

    std::cout << myB->getAVec()[0].getData() << std::endl; // prints "0" as expected

    auto pyModule = pybind11::module::import("mod");
    pyModule.attr("myB") = pybind11::cast(myB);

    pybind11::object scope = pybind11::module::import("__main__").attr("__dict__");
    pybind11::eval<pybind11::eval_mode::eval_single_statement>("import mod", scope);
    pybind11::eval<pybind11::eval_mode::eval_single_statement>("mod.myB.get_A_vec()[0].set_data(1)", scope);

    std::cout << myB->getAVec()[0].getData() << std::endl; // prints "0" after #1455 merged "1" before

    return 0;
}

```

It would appear that a copy is made of the vector in the example, instead of passing an opaque vector as was previously the case.

**Comment by clauniel** [comment]
So this example is working and I cannot figure out why. I'm not able to reproduce my issue.

In the end I made the relevant types opaque and that seemed to make pybind happy. I don't need to to that for my minimal example though. I don't see why it works. 

Anyway, there was some change to the way containers must be bound in order to actually return references.

**Comment by EricCousineau-TRI** [comment]
As a follow-up, I'm also apparently affected by a variant of this:
https://github.com/RobotLocomotion/drake/pull/9368

This is different, in that I am not attempting to make `std::vector<T>` opaque, but would rather prefer to get a copy.
However, I'm trying to make a simplified reproduction off of `master`, and am unable to reproduce the issue. I am using a fork of `pybind`, but my assumption is that this error should be invariant of any changes I've made (maybe?).

**Comment by EricCousineau-TRI** [comment]
@wjakob I believe this is a duplicate of #1561, resolved by #1603, and perhaps could be closed?

Using the latest `master` in our fork resolves the above issue I mentioned.

**Comment by wjakob** [comment]
Okay, thanks for letting me know!
